File: frontend/src/lib/validations/hapticConfig.ts
import { z } from 'zod';

// Android haptic settings
const androidHapticSchema = z.object({
  duration: z.number({ required_error: 'Duration is required' }).int().min(0),
  amplitude: z.number({ required_error: 'Amplitude is required' }).int().min(0).max(255),
});

// iOS haptic settings
const iosHapticSchema = z.object({
  intensity: z.number({ required_error: 'Intensity is required' }).min(0).max(1),
  sharpness: z.number({ required_error: 'Sharpness is required' }).min(0).max(1),
  duration: z.number({ required_error: 'Duration is required' }).min(0),
});

// Haptic type (contains both Android and iOS settings)
const hapticTypeSchema = z.object({
  android: androidHapticSchema,
  ios: iosHapticSchema,
});

// Full haptic config
export const hapticConfigSchema = z.object({
  soft: hapticTypeSchema,
  light: hapticTypeSchema,
  medium: hapticTypeSchema,
  heavy: hapticTypeSchema,
  button: hapticTypeSchema,
  success: hapticTypeSchema,
  error: hapticTypeSchema,
});

// Type exports
export type AndroidHaptic = z.infer<typeof androidHapticSchema>;
export type IOSHaptic = z.infer<typeof iosHapticSchema>;
export type HapticType = z.infer<typeof hapticTypeSchema>;
export type HapticConfig = z.infer<typeof hapticConfigSchema>;

const defaultHapticType: HapticType = {
  android: { duration: 0, amplitude: 0 },
  ios: { intensity: 0, sharpness: 0, duration: 0 },
};

export const defaultHapticConfig: HapticConfig = {
  soft: { ...defaultHapticType },
  light: { ...defaultHapticType },
  medium: { ...defaultHapticType },
  heavy: { ...defaultHapticType },
  button: { ...defaultHapticType },
  success: { ...defaultHapticType },
  error: { ...defaultHapticType },
};

File: backend/app/schemas/config_sections/haptic_config.py
from pydantic import BaseModel, Field
from typing import Optional


class AndroidHaptic(BaseModel):
    """Android-specific haptic settings"""
    duration: int = Field(..., ge=0, description="Duration in milliseconds")
    amplitude: int = Field(..., ge=0, le=255, description="Amplitude (0-255)")
    


class IOSHaptic(BaseModel):
    """iOS-specific haptic settings"""
    intensity: float = Field(..., ge=0, le=1, description="Intensity (0.0-1.0)")
    sharpness: float = Field(..., ge=0, le=1, description="Sharpness (0.0-1.0)")
    duration: float = Field(..., ge=0, description="Duration in seconds")
    


class HapticType(BaseModel):
    """Haptic settings for a specific feedback type with platform-specific configs"""
    android: AndroidHaptic = Field(..., description="Android haptic settings")
    ios: IOSHaptic = Field(..., description="iOS haptic settings")
    


class HapticConfig(BaseModel):
    """Full haptic configuration with all feedback types"""
    soft: HapticType = Field(..., description="Soft haptic feedback")
    light: HapticType = Field(..., description="Light haptic feedback")
    medium: HapticType = Field(..., description="Medium haptic feedback")
    heavy: HapticType = Field(..., description="Heavy haptic feedback")
    button: HapticType = Field(..., description="Button haptic feedback")
    success: HapticType = Field(..., description="Success haptic feedback")
    error: HapticType = Field(..., description="Error haptic feedback")
    



